### DISCUSSION TITLE
Unable to download file using GetMalQueryDownloadV1

### DISCUSSION THREAD

**Discussion by MonalisaBhattacharya** [CrowdStrike/falconpy#304] [Q&A] [answered]
I tried using the below script to download a file 
https://github.com/CrowdStrike/falconpy/blob/main/docs/service-collections/malquery.md#GetMalQueryDownloadV1
from falconpy import api_complete as FalconSDK

falcon = FalconSDK.APIHarness(creds={
      'client_id': falcon_client_id,
      'client_secret': falcon_client_secret
   }
)

IDS = 'ID1,ID2,ID3'

response = falcon.command('GetMalQueryDownloadV1', ids=IDS)
print(response)
falcon.deauthenticate()

But I am getting the below output
![image](https://user-images.githubusercontent.com/11628577/129676587-3187c9e5-ed19-4935-b405-ea8753e0e218.png)

I am new to python. Could you tell me where i am going wrong..

**Comment by jshcodes** [comment]
Hi @MonalisaBhattacharya!

You're getting the actual file as a response, as opposed to a dictionary (for our purposes, a JSON formatted result). This is expected with this operation, so we need to adjust the code to store the response to a file.

Try this and let us know if you have issues.
```python3
from falconpy import api_complete as FalconSDK

falcon = FalconSDK.APIHarness(creds={
        'client_id': "YOUR_CLIENT_ID_HERE",
        'client_secret': "YOUR_CLIENT_SECRET_HERE"
        }
    )

id_to_retrieve = "ID_GOES_HERE"
save_file = "filename.ext"
response = falcon.command('GetMalQueryDownloadV1', ids=id_to_retrieve)
open(save_file, 'wb').write(response)
falcon.deauthenticate()
```

> Side note: The wiki is in the process of being refreshed. We will make sure the help content for this gets updated. Thank you for pointing it out. :smile:

**Comment by MonalisaBhattacharya** [comment]
Thanks CrowdStrike/falconpy,

I am able to download the file. It is stored as filename.text. How would I
be able to understand the file type from the hash if all the files will be
stored as text file. Just curious.

Thanks
Monalisa

**Comment by jshcodes** [comment]
The file name is arbitrary in this example, you can name it whatever you want. From the above I'd guess it was a zip file, but that's just a guess.  (Typically, you pull the name at the same time you pull the ID and use that for the save_file value.)

**Comment by jshcodes** [comment, marked as answer]
Hi @MonalisaBhattacharya -

Give this a shot. You may have to adjust things a bit to suit your purposes.

```python3
from falconpy import api_complete as FalconSDK

falcon = FalconSDK.APIHarness(creds={                   # Authenticate
        'client_id': "YOUR_CLIENT_ID_HERE",
        'client_secret': "YOUR_CLIENT_SECRET_HERE"
        }
    )

malware = falcon.command(                               # This command will search
    "PostMalQueryFuzzySearchV1",                        # MalQuery for a reference to
    body={                                              # "trickbot", may have to adjust
        "options": {
            "limit": 1                                  # I limited the results to 1
        },                                              # for example purposes
        "patterns": [{
            "type": "ascii",
            "value": "trickbot"                         # ASCII pattern: `trickbot`
            }]
        }
    )["body"]["resources"][0]                           # Retrieve the first resource returned

id_to_retrieve = malware["sha256"]                      # Retrieve the ID of the malware
search_request_id = falcon.command(                     # Request the download
        "PostMalQueryEntitiesSamplesMultidownloadV1",
        body={
            "samples": [id_to_retrieve]
        }
    )["body"]["meta"]["reqid"]

running = True                                          # Boolean tracking our running wait status
while running:
    search_result = falcon.command(
        "GetMalQueryRequestV1",
        ids=search_request_id
    )
    status = search_result["body"]["meta"]["status"]    # Grab our status result
    if status == "done":                                # When it's done, continue
        running = False

save_file = "infected.zip"                              # Save file name
archive_result = falcon.command(                        # Request the download
    "GetMalQueryEntitiesSamplesFetchV1",
    ids=search_request_id
)
if isinstance(archive_result, dict):                    # Is it a dictionary?
    print(                                              # This probably means there was an error
        archive_result["body"]["errors"][0]["message"]  # Show the error message we got back
        )
else:                                                   # No, it's a file
    open(save_file, 'wb').write(archive_result)         # Write the response to our save file

falcon.deauthenticate()                                 # Deauthenticate
```

**Reply by MonalisaBhattacharya** [reply]
Thanks.. Appreciate your help..
